File: python/0102_binary-tree-level-order-traversal.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from collections import deque

class Solution(object):
    def levelOrder(self, root: TreeNode) -> list[list[int]]:
        if not root:
            return []
        current_queue = deque()
        next_queue = deque()
        current_queue.append(root)
        result = []

        while current_queue:
            result.append([n.val for n in current_queue])
            while current_queue:
                node = current_queue.popleft()
                if node.left:
                    next_queue.append(node.left)
                if node.right:
                    next_queue.append(node.right)
            current_queue = next_queue
            next_queue = deque()

        return result

File: src/bin/0102-binary-tree-level-order-traversal.rs
// @leetup=info id=102 lang=rust slug=binary-tree-level-order-traversal


// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @leetup=code
use std::cell::RefCell;
use std::collections::*;
use std::rc::Rc;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root == None {
            return vec![];
        }
        let mut cur_q = VecDeque::new();
        cur_q.push_back(root.unwrap());
        let mut next_q = VecDeque::new();
        let mut result: Vec<Vec<i32>> = Vec::new();

        while !cur_q.is_empty() {
            result.push(cur_q.iter().map(|n| n.borrow().val).collect());
            while let Some(n) = cur_q.pop_front() {
                if let Some(n) = &n.borrow().left {
                    next_q.push_back(n.clone());
                }
                if let Some(n) = &n.borrow().right {
                    next_q.push_back(n.clone());
                }
            }
            std::mem::swap(&mut cur_q, &mut next_q);
        }
        result
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::level_order;
}
